### PR TITLE
Search with pagination

### PR DESCRIPTION
**Pull request by hedgehog1833**
yeha, first one \o/
general search and pagination

### CODE REVIEW

File: src/main/java/com/metalr2/config/mvc/WebConfig.java [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
   @Override
   public void addViewControllers(ViewControllerRegistry registry) {
-    registry.addViewController(Endpoints.Frontend.FOLLOW_ARTISTS).setViewName(ViewNames.Frontend.FOLLOW_ARTISTS);
+//    registry.addViewController(Endpoints.Frontend.FOLLOW_ARTISTS).setViewName(ViewNames.Frontend.FOLLOW_ARTISTS);
```
Comment (Waginator) [inline src/main/java/com/metalr2/config/mvc/WebConfig.java]: remove comment

File: src/main/java/com/metalr2/web/controller/SearchController.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Controller
+@Slf4j
+public class SearchController {
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchController.java]: the name of controller should be FollowArtistsController

File: src/main/java/com/metalr2/web/controller/SearchController.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  @ModelAttribute
+  private SearchRequest searchRequest() {
+    return new SearchRequest();
+  }
+
+  @PostMapping({Endpoints.Frontend.FOLLOW_ARTISTS})
+  public ModelAndView handleSearchRequest(@ModelAttribute SearchRequest searchRequest) {
+    log.info(searchRequest.getArtistName());
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchController.java]: Logging should be more descriptive here. If the application is running productively, it is not very meaningful if only the name of the artists appears in the log file. Log-Level is in my opinion DEBUG or TRACE.

Furthermore: If the log output gets values from a method call, then the following procedure is best practice: log.debug("{}", searchRequest.getArtistName());

Reason: The method is now called only if the log level of the application is DEBUG. The reason for this is performance. Possibly there is an expensive operation behind the method call.

File: src/main/java/com/metalr2/web/controller/SearchController.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  @GetMapping({Endpoints.Frontend.FOLLOW_ARTISTS})
+  public ModelAndView showSearch(@RequestParam(name = "size", defaultValue = DEFAULT_PAGE_SIZE) int size,
+                                 @RequestParam(name = "page", defaultValue = DEFAULT_PAGE) int page,
+                                 @RequestParam(name = "artistName", defaultValue = DEFAULT_ARTIST_NAME) String artistName) {
+    if (artistName.isEmpty()) {
+      return getDefaultModelAndView();
+    }
+
+    Optional<ArtistSearchResults> artistSearchResultsOptional = artistSearchRestClient.searchForArtist(artistName, String.valueOf(page), String.valueOf(size));
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchController.java]: Please compare lines 49 to 60 and 71 to 81. I think you can extract this into a method.

File: src/main/java/com/metalr2/web/controller/SearchController.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    return new ModelAndView(ViewNames.Frontend.FOLLOW_ARTISTS, viewModel);
+  }
+
+  private Map<String,Object> buildViewModel(ArtistSearchResults artistSearchResults, String artistName) {
+    Map<String, Object> viewModel = new HashMap<>();
+    viewModel.put("artistName", artistName);
+    viewModel.put("artistSearchResultList", artistSearchResults.getResults());
+
+    Pagination pagination = artistSearchResults.getPagination();
+    List<String> pageNumbers = IntStream.rangeClosed(1, pagination.getPagesTotal()).boxed().map(String::valueOf).collect(Collectors.toList());
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchController.java]: is transformation to String really necessary?

File: src/main/java/com/metalr2/web/controller/SearchController.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    viewModel.put("currentPage", pagination.getCurrentPage());
+    viewModel.put("pageNumbers", pageNumbers);
+
+    PaginationUrls paginationUrls = pagination.getUrls();
+    String nextSize = DEFAULT_PAGE_SIZE;
+    String nextPage = DEFAULT_PAGE;
+
+    if (paginationUrls.getNext() != null){
+      nextSize = String.valueOf(pagination.getItemsPerPage());
+      nextPage = String.valueOf(pagination.getCurrentPage()+1);
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchController.java]: I think you don't need to convert the values into a String

File: src/main/java/com/metalr2/web/dto/request/SearchRequest.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class SearchRequest {
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/dto/request/SearchRequest.java]: the name should be ArtistSearchRequest

File: src/main/resources/static/css/frontend.css [hunk 8]
```diff
@@ -47,3 +47,28 @@ footer{
...
+.btn-pagination {
+    margin-right: 0.5rem;
+    background-color: #ececec;
+    color: #212529;
+    padding: 0.25rem;
+}
+
+.btn-pagination:hover {
+    background-color: #777777;
+}
```
Comment (Waginator) [inline src/main/resources/static/css/frontend.css]: add new line

File: src/main/resources/templates/frontend/follow-artists.html [hunk 9]
```diff
@@ -1,6 +1,55 @@
 <!DOCTYPE html>
 <html data-th-replace="~{frontend/fragments/layout :: layout (~{::body})}" lang="en">
-  <body>
+<body>
```
Comment (Waginator) [inline src/main/resources/templates/frontend/follow-artists.html]: The body and the first div container are not correctly indented...

Should be:
```
<html>
  <body>
    <div>
    ...
    </div>
  </body>
</html>
```

File: src/main/java/com/metalr2/web/controller/SearchController.java (hunk 6 above)
Comment (hedgehog1833) [reply src/main/java/com/metalr2/web/controller/SearchController.java]: Not there specifically, that's correct. But there are some places where I need the String and some where I need an int.
I tried to improve it somewhat but it's still kind of weird, if you don't want to hold both types independantly

File: src/main/resources/templates/frontend/follow-artists.html (hunk 9 above)
Comment (hedgehog1833) [reply src/main/resources/templates/frontend/follow-artists.html]: IntelliJ suggested this for me, but I guess your version looks better

**Pull request merged**